lib/ukprint/logbuf.c

```diff
@@ -0,0 +1,144 @@
+/* SPDX-License-Identifier: BSD-3-Clause */
+/* Copyright (c) 2024, Unikraft GmbH and The Unikraft Authors.
+ * Licensed under the BSD-3-Clause License (the "License").
+ * You may not use this file except in compliance with the License.
+ */
+
+#include <errno.h>
+#include <string.h>
+
+#include <uk/assert.h>
+#include <uk/console.h>
+#include <uk/spinlock.h>
+
+#include "snprintf.h"
+#include "uk_print_msg.h"
+
+#define LOGMSG_ALIGN		16
+#define LOGMSG_SIZE(_msg_len)	(sizeof(struct uk_print_logmsg) + (_msg_len))
+
+struct uk_print_logmsg {
+	struct uk_print_logmsg *next;
+	struct uk_print_msg msg;
+};
+
+static __u8 logbuf[CONFIG_LIBUKPRINT_LOGBUF_SIZE];
+
+static struct uk_print_logmsg *logbuf_head;
+static struct uk_print_logmsg *logbuf_tail;
+
+static uk_spinlock logbuf_lock = UK_SPINLOCK_INITIALIZER();
+
+/* Returns the next free aligned position in the buffer, or wraps
+ * around.
+ */
+static inline __uptr logbuf_next(void)
+{
+	__uptr next;
+
+	if (!logbuf_head)
+		return (__uptr)logbuf;
+
+	next = ALIGN_UP((__uptr)logbuf_tail + LOGMSG_SIZE(logbuf_tail->msg.len),
+			LOGMSG_ALIGN);
+
+	if (next >= (__uptr)logbuf + sizeof(logbuf))
+		next = (__uptr)logbuf;
+
+	return next;
+}
+
+static inline void update_head(struct uk_print_logmsg *logmsg, __sz msg_len)
+{
+	struct uk_print_logmsg *lm;
+	int found = 0; /* bool */
+
+	if (RANGE_OVERLAP((__uptr)logbuf_head, LOGMSG_SIZE(logbuf_head->msg.len),
+			  (__uptr)logmsg, LOGMSG_SIZE(msg_len))) {
+		for (lm = logbuf_head; lm != lm->next; lm = lm->next) {
+			if ((__uptr)lm->next > (__uptr)logmsg + LOGMSG_SIZE(msg_len)) {
+				logbuf_head = lm->next;
+				found = 1;
+				break;
+			}
+		}
+
+		if (!found) {
+			/* This logmsg is the last element in the buffer.
+			 * Set head to the first element.
+			 */
+			logbuf_head = (struct uk_print_logmsg *)logbuf;
+		}
+	}
+}
+
+int uk_print_logbuf_write(struct uk_print_msg *msg)
+{
+	struct uk_print_logmsg *logmsg;
+	va_list cp;
+
+	uk_spin_lock(&logbuf_lock);
+
+	va_copy(cp, msg->ap);
+	msg->len = vsnprintf(msg->msg, 0, msg->fmt, cp);
```

Comment:
I see that `msg->msg` is a flexible array but all the places that pass the `uk_print_msg ` argument to this function declare it as a plain stack local variable. Is it really safe to write to `msg->msg` here? Or maybe I am missing something.